514 515 render time bound changes home and ft

themagicianking opened the pull request:
Issue: Changing the website areas twice a year with the same text and links feels redundant. Can this be automated to render content based on the times / dates we specify? (#514 and #515)

Review request: Double check that logic works correctly, advise whether "extended" message should display before or after original closing date (currently displays as soon as extended variable is changed to true)

Testing and changes: This change requires the addition of the variables APP_OPEN_DATE and IS_EXTENDED to the env file. Below is an example file:
```
APP_OPEN_DATE = '01/26/24 12:00:00'
APP_EXTENDED = 'true'
```
APP_OPEN_DATE should be in a format that can be parsed by [the Python datetime library's strptime method](https://docs.python.org/3/library/datetime.html#datetime.datetime.strptime). APP_EXTENDED takes a string version of a boolean (case insensitive).

To test: Change the variables APP_OPEN_DATE and APP_EXTENDED to test. You should notice the following:

If today's date is more than 28 days past APP_OPEN_DATE and APP_EXTENDED is false or today's date is more than 42 days past APP_OPEN_DATE and APP_EXTENDED is true:

- The first button on the home page over the banner should say "Outsource your software work!" and link to the sponsor application page
- The first button on the full time program page should say "Fill out our interest form to be notified about our next cohort!" and link to our interest form
- The second button on the full time program page should say "Sign up to join our events" and link to our Eventbrite sign up

If today's date is less than 28 days past APP_OPEN_DATE and APP_EXTENDED is false:

- The first button on the home page over the banner should say "Apply by (date 28 days from APP_OPEN_DATE)!" and link to the full time program page
- The first button and the second button on on the full time program page should say "Apply by (date 28 days from APP_OPEN_DATE)!" and link to our application form

If today's date is less than 42 days past APP_OPEN_DATE and APP_EXTENDED is true:
- The first button on the home page over the banner should say "Extended! Apply by (date 28 days from APP_OPEN_DATE)!" and link to the full time program page
- The first button and the second button on on the full time program page should say "Extended! Apply by (date 28 days from APP_OPEN_DATE)!" and link to our application form

Tasks done in this PR:

- [x] Created the following environment variables (to be added to the .env file): APP_OPEN_DATE, IS_EXTENDED
- [x] Created a helper function get_time() in main_site.py that uses the added environment variables to return a dictionary containing a boolean for if applications are open, and the text to be added to buttons on the homepage and full time program page if applications are open. The text will include whether or not applications have been extended.
- [x] Refactored buttons in home.html and full-time-program.html to render text and links based on the information passed in by the times parameter (which is created with get_time())
- [x] Made some formatting changes to old code to correct flake8 violations

New helper function is as follows:
```
def get_time():
    app_open_date_string = os.environ["APP_OPEN_DATE"]
    is_extended = os.environ["APP_EXTENDED"].lower() == "true"
    app_open_date = datetime.datetime.strptime(
        app_open_date_string, "%m/%d/%y %H:%M:%S"
    )
    today = datetime.datetime.today()
    if is_extended:
        app_close_date = app_open_date + datetime.timedelta(days=42)
        date_string = app_close_date.strftime("%B %-d")
        text = "Extended! Apply by {date} (12pm PT)!".format(date=date_string)
    else:
        app_close_date = app_open_date + datetime.timedelta(days=28)
        date_string = app_close_date.strftime("%B %-d")
        text = "Apply by {date} (12pm PT)!".format(date=date_string)
    app_open = app_open_date <= today <= app_close_date

    return {
        "app_open": app_open,
        "text": text,
    }
```

Notes:
- I created another pr [here](https://github.com/Techtonica/techtonica.org/pull/526) for the parent issue, #457, but it was created incorrectly with some unwanted commits so I will delete it as soon as I have time. It also does not deal with the other child issue so it feels more appropriate to make a separate pr. 
- This PR was coauthored by @jsnorek (please add any additional comments you have :) )
- Dates are calculated as follows: it is decided that applications are closed if it has been 28 days from the open date and applications are **not* extended, or if it has been 42 days from the open date and applications **are** extended. Applications are open if the above criteria have not been met and they are extended if  IS_EXTENDED is true, regardless of whether or not things are still within the original open window.

jsnorek commented:
@themagicianking I'm having trouble running this branch because of the missing variables and .env file - I'm thinking we need to include a sample .env file based on the one in your local!

themagicianking commented:
@jsnorek I added a sample .env and instructions for how to fill in the values :)

themagicianking commented:
@jsnorek Changed env variables to config variables!

jsnorek commented:
Awesome, the site is now running locally for me! Buttons seem to work great!!

daaimah123 commented on main_site.py:
```diff
@@ -178,7 +186,10 @@ def render_ft_program_page():
     """
     Renders the full-time program page from jinja2 template
```
this comment should share a few words more about what is happening before the render of the FT page

themagicianking replied:
Resolved in [this commit](https://github.com/Techtonica/techtonica.org/pull/534/commits/e980faecb1b9eeced035a487599a02c2845cea3f)!

daaimah123 commented on main_site.py:
```diff
@@ -226,6 +240,28 @@ def get_events():
         return []
 
 
+def get_time(app_open_date_string, is_extended_string):
```
please provide a brief comment sharing what this function does and how it relates to the other templates/files

themagicianking replied:
Resolved in [this commit!](https://github.com/Techtonica/techtonica.org/pull/534/commits/f26da869aa7571bca0ae619f9c165bd63a267fc4)

daaimah123 commented on templates/team.html:
```diff
@@ -23,15 +23,23 @@ <h1>Staff</h1>
-        for underrepresented groups in tech. Daaimah grew up and lives in the
-        Bay Area. She enjoys hiking with her family and learning about science
-        and anthropology.
+        As a queer, black, disabled, primary caregiver, Daaimah faced many
+        barriers simply entering the tech industry and when getting into her
+        first technical role; she discovered it was a whole new beast compared
+        to her past work in the grassroots nonprofit world: writing grants,
+        managing programs, supporting equity and inclusion through public policy
+        education, and advising persons with disabilities on careers. Following
+        completion of the Techtonica program in 2019, she worked at SurveyMonkey
+        as a software engineer, tech lead at Hack The Hood, and technical career
+        counselor with Code Tenderloin. Wanting to pay it forward, Daaimah
+        volunteered as a graduate board member and mentor program manager at
+        Techtonica until she was hired as the Software Training Engineering
+        Manager at Techtonica, where she now utilizes the skill sets in her
+        arsenal to advocate for mindful accommodations, training access, and new
+        pathways to entry for underrepresented intersectionalities in the tech
+        space. She grew up and still resides in the Bay Area. For kicks, she
+        enjoys hitting the trails with her two children and two dogs, as well as
+        learning about science and anthropology.
```
Please see suggestions for what is currently live:
```suggestion
        Daaimah, a queer, black, disabled individual and primary caregiver, entered the tech industry facing numerous barriers. After completing the Techtonica program in 2019, she worked as a software engineer at SurveyMonkey, tech lead at Hack The Hood, and technical career counselor with Code Tenderloin. She later became the Software Training Engineering Manager at Techtonica, advocating for accommodations and training access for underrepresented groups in tech. Daaimah grew up and lives in the Bay Area. She enjoys hiking with her family and learning about science and anthropology.
```

themagicianking replied:
Resolved in new rebase!

daaimah123 commented on templates/team.html:
```diff
@@ -100,6 +78,8 @@ <h1>Staff</h1>
           rel="noopener noreferrer"
           >Margaret Fero (they/them), <br />Sponsorships Lead</a
         >
+          >Margaret Fero (they/them), <br />Sponsorships Lead</a
```
I wonder if you are having some rebasing issues, this is added twice it seems; please remove this duplication and resolve any rebasing or merging conflicts

themagicianking replied:
I believe I had mistakenly originally branched this from main instead of develop and it led to some duplication when I opened things in my merge editor. This should be fixed in my most recent rebase.

daaimah123 commented on templates/team.html:
```diff
@@ -127,20 +114,87 @@ <h1>Staff</h1>
         >
       </footer>
       <p class="justify">
+        The need for more diversity, inclusion, and equity in tech and the
```
another duplication

daaimah123 commented on templates/team.html:
```diff
@@ -110,10 +90,17 @@ <h1>Staff</h1>
         not be represented in key conversations about technology. In their spare
         time, Margaret enjoys basking in the many useful services of local
         libraries.
+        Margaret Fero is an interdisciplinary hacker with professional
+        experience in education, technical writing, software development, and
+        information security. They are excited to apply their background to
+        share Techtonica's mission with companies and individuals who share our
+        desire to economically empower people whose perspectives may otherwise
+        not be represented in key conversations about technology. In their spare
+        time, Margaret enjoys basking in the many useful services of local
+        libraries.
```
another duplication

daaimah123 commented on templates/team.html:
```diff
@@ -127,20 +114,87 @@ <h1>Staff</h1>
           target="_blank"
           rel="noopener noreferrer"
           >Michelle Glauser (she/her), <br />Founder and CEO</a
+          >Michelle Glauser (she/her), <br />Founder and CEO</a
```
another duplication

daaimah123 commented on templates/team.html:
```diff
@@ -127,20 +114,87 @@ <h1>Staff</h1>
         #ILookLikeAnEngineer ad campaign, and Techtonica. Michelle has also
         worked as a developer advocate. She loves biking, reading, making and
         eating cookies, and going on walks with her fluffy dog.
+        to organize support groups, meetups like PyLadies SF (the world's
+        largest chapter of women who code in Python), workshops, the
+        #ILookLikeAnEngineer ad campaign, and Techtonica. Michelle has also
+        worked as a developer advocate. She loves biking, reading, making and
+        eating cookies, and going on walks with her fluffy dog.
```
another duplication

daaimah123 commented on templates/team.html:
```diff
@@ -127,20 +114,87 @@ <h1>Staff</h1>
+        <p>
+          Alexandria has worked in tech for 13 years — at Yahoo, Facebook,
+          LinkedIn, Airbnb, The Chan Zuckerberg Initiative, and currently Lyft.
+          Alexandria has carried roles of Recruiter + D&I thought leader to
+          elevate programs and partnerships, along with re-designing approaches
+          to hiring. Beyond her day-to-day, she does her knowledge-sharing
+          through think tank sessions with various nonprofits and panels such as
+          Black Girls Code, Grace Hopper Conference, and Expat Women. She is
+          passionate about diversity engagement and community building.
+        </p>
+      </div>
+      <div class="column centered">
+        <img
+          src="{{ url_for('static', filename='img/Annie-Tsai-min.jpg') }}"
+          alt="Annie Tsai"
+          class="resize-200px"
+        />
+        <footer>
+          <a
+            href="https://www.linkedin.com/in/meannietsai/"
+            target="_blank"
+            rel="noopener noreferrer"
+            >Annie Tsai (she/her)</a
+          >
+        </footer>
+        <p>
+          Annie Tsai is the Chief Operating Officer at Interact, a SaaS
+          marketing software company, and part-time Venture Partner at The House
+          Fund. A graduate of the University of California at Berkeley (2001 —
+          Go Bears!), Annie has 20+ years in building, growing, and leading a
+          variety of functions across the non-technology side of the house at
+          software companies in the US, EMEA, and APAC. Annie's passions include
+          lifting up working mothers, ensuring the products are built by people
+          who look like their customers, investing in improving public
+          education, and ensuring every child has a safe place to sleep at
+          night. She lives in the Bay Area with her husband, two children, and
+          pup.
+        </p>
+      </div>
+    </div>
```
another duplicate

daaimah123 commented on templates/team.html:
```diff
@@ -326,6 +447,9 @@ <h1 class="blue-h1">Board Members</h1>
           mountains.
         </p>
       </div>
+      <div class="column centered">
+        <p>Thane Wilson :)</p>
+      </div>
```
Yep, definitely some older changes here. Aw remember your very first commit, @themagicianking!

daaimah123 reviewed (changes requested):
A rebase is needed to get the most up to date info from develop into this branch

themagicianking commented:
@daaimah123 rebase should be completed (correctly this time without duplications :) ) and everything else has been resolved!
@monikkaelyse rebase is done!

themagicianking commented:
@monikkaelyse Gotcha, thanks for the heads up about the ENV file!

I fixed the merge conflict code [here](https://github.com/Techtonica/techtonica.org/pull/534/commits/c1f6e9a770c4c75ff7b9b9417c81f880ae022c15), I honestly have no idea when it was introduced either haha

jsnorek commented:
@monikkaelyse 

I believe both sets of variables work the same (with respectively different button displays) but for example, 

`sponsor_open_date = 01/20/24 13:55:26 `- is the date we set the sponsor applications open.
`sponsor_extend = true` accounts for whether we are extending that application or not (true or false)

`get_time()` in main_site.py is built to extend the application dates by 14 days if `sponsor_extend = true`.
It's also built to calculate the close date (if `is_extended` it adds 42 days from the `sponsor_open_date`, otherwise it adds 28)

If the calculated `app_close_date` of the application hasn't passed yet then

if `sponsor_extend = true`, the displayed date on the button is updated to show the extended deadline date.
if it's false, it will display the `app_close_date` from `get_time()`

If the closed date has passed today's date, the button will display "Outsource your software work!"

So with these variables:
```
sponsor_open_date = 01/20/25 13:55:26
sponsor_extend = false
```

the **expected result** is the button on the home page displaying "Apply by February 17 (12pm PT)" and it links to the full-time program page

with:
```
sponsor_open_date = 01/02/25 13:55:26
sponsor_extend = true
```

**expected result** is "Extended! Apply by February 13 (12pm PT)" and still links to full-time program page

with
```
sponsor_open_date = 01/02/25 13:55:26
sponsor_extend = false
```

**expected result** is "Outsource your software work!" and links to the sponsor page

themagicianking commented:
@jsnorek That explanation is perfect, thank you :)

themagicianking commented:
@monikkaelyse I'm investigating the 404 issue now, I have had a lot of trouble rebasing

jsnorek commented:
@monikkaelyse looking back, my ticket didn't specify what the dates were in reference to so I thought it was a sponsor application date, especially since there was another button ticket for the ft applications. But if there are no sponsor applications, it would make sense for there to only be one variable. 

@themagicianking I think we only need one variable then

themagicianking commented:
@monikkaelyse I moved the new variables back to the env file and updated the code to match. The code now uses only two environment variables, just for application open date/extension. I also added a note clarifying exactly how the function calculates dates and generates button text, in case any changes are wanted in relation to that. @jsnorek brought up that as of right now, if IS_EXTENDED is true, the button will display "Extended!", whether or not the current date is in the original application window. Neither of us were sure if this is desired behavior or not.

Other than that there are no other changes :) @daaimah123 if everything looks good I can go ahead and merge this in.

themagicianking commented:
@monikkaelyse absolutely! The edit I made details how to test the code and should hopefully make the changes we made clearer as well.

With regards to your second comment: what I meant is that if we're inside of the original application window, the button will say extended. Eg let's say that today is May 10th and the normal day for applications to close is May 14th. Currently, the button text should you the generic "Apply by May 14th!" If I change the environment variable IS_EXTENDED to true on May 10th, the button will say "Extended! Apply by May 28th!". Our uncertainty was in whether or not the button should continue to say "Apply by May 14th!" until May 15th, and then change to the extended message.

themagicianking commented:
@monikkaelyse Thank you so much! I'd combed back through the code to try and find where the image went missing, and I thought I'd fixed the line that caused it but I double checked in my local copy and it's still not displaying. I agree that was probably a rebasing issue but I am also not sure quite how to fix it. I took a look at the files changed and I think the pre commit yaml and base.html should not be in there; I assume those got caught in the mix during rebasing. What would be the best way to fix this moving forward?

themagicianking commented:
@monikkaelyse I reverted all the files that were accidentally altered in the rebase from the commit to their current state in develop. The image on the apply page is working for me locally--can you confirm that it's working for you as well? Thanks!

themagicianking commented:
@monikkaelyse Can you point me to which changes you're referring to? As of right now, looking over the difference, I don't see any not made by me or @jsnorek, although I do see quite a few that were made by either prettier or black in pre commit. Should those be removed?

themagicianking commented:
@monikkaelyse I'm unsure if this is because I'm on mobile but for some reason when I try to look at your review I'm simply redirected to the commit history. I thought this was because one of the changes you requested was resolved (having multiple variables for the open date) but I'll take another look at it when I'm back in office and can go over everything properly. Thank you so much for your patience! I really appreciate it as well as all the help you've given us in reviewing this PR  :)

jsnorek commented:
@themagicianking @monikkaelyse I am also not seeing the unresolved comments! Anything I click either brings me to Monikka's original request/question with the extra date variable or it shows this
![Screenshot 2025-02-05 at 4 37 04 PM](https://github.com/user-attachments/assets/bb908350-3c18-4287-b113-c3d3b9287b85)


When I look at the files changed page I don't see any comments there either.

themagicianking commented:
@monikkaelyse I can see all your requested changes! I'm looking over them now, weird that I just could not see them before, I wonder if it's because the line numbers changed...? In any case, I'll review them and get back to you when I've done so.

themagicianking commented on main_site.py:
```diff
@@ -226,6 +237,30 @@ def get_events():
 
+# returns string & extension variable to display time bound page components
+def get_time():
+    app_open_date_string = os.environ["APP_OPEN_DATE"]
+    is_extended = os.environ["APP_EXTENDED"].lower() == "true"
```
Got it! So, essentially something similar to the try catch block that's executed if the eventbrite token is not set--I'll go ahead and add that.

themagicianking replied:
Resolved [here](https://github.com/Techtonica/techtonica.org/pull/534/commits/dd9762a9b54cac0f67360836e7a5da5a1bd5dc09) :)

themagicianking commented on main_site.py:
```diff
@@ -226,6 +237,30 @@ def get_events():
+    is_extended = os.environ["APP_EXTENDED"].lower() == "true"
+    app_open_date = datetime.datetime.strptime(
+        app_open_date_string, "%m/%d/%y %H:%M:%S"
+    )
```
Resolved [here](https://github.com/Techtonica/techtonica.org/pull/534/commits/797ed5220eb2a4bd2ddb3f695b5f6ab09c1c6623)!

themagicianking commented on main_site.py:
```diff
@@ -226,6 +237,30 @@ def get_events():
 
 
+# returns string & extension variable to display time bound page components
+def get_time():
```
Resolved [here](https://github.com/Techtonica/techtonica.org/pull/534/commits/970e2bca56da73e8210180c9eaed2e2dfc72e027)!

themagicianking commented on static/css/style.css:
```diff
@@ -270,7 +269,8 @@ button {
   height: 100vh;
-  background: url(../img/2024-H2-Launch-Celebration-min.png) #0093b5 no-repeat top fixed;
+  background: url(../img/2022-H1-Launch-Celebration.png) #0093b5 no-repeat top
+    fixed;
```
Outdated 👍

themagicianking commented on templates/base.html:
```diff
@@ -568,123 +568,119 @@
     </style>
```
Also outdated, I do believe this was prettier, I've removed this change from the PR.

themagicianking commented on templates/full-time-program.html:
```diff
@@ -5,238 +5,311 @@
+        Hands on software engineering training with need-based, sliding-scale,
+        subsidized tuition and stipend scholarships, followed by six months of
+        remote job placement or job search support for women and non-binary
+        adults seeking economic empowerment. The 2025 H1 Cohort kicks off in
+        January.
```
Would you suggest simply deleting it or changing it to past tense?

themagicianking commented on templates/full-time-program.html:
```diff
@@ -5,238 +5,311 @@
+        adults seeking economic empowerment. The 2025 H1 Cohort kicks off in
+        January.
       </p>
       <!-- <iframe src="https://docs.google.com/forms/d/e/1FAIpQLSdk8nJUSuK_xoILyYyf3GIpVypQRtqsx9aQE7odHgX1cWvoHA/viewform?embedded=true" width="640" height="600" frameborder="0" marginheight="0" marginwidth="0">Loading…</iframe>       -->
```
Same question as earlier for this; should it be deleted or changed to past tense?

themagicianking replied:
Oh gotcha! I'm unsure; it was commented out when I received the ticket.

jsnorek replied:
@monikkaelyse since it was already commented out when we started, do you want us to delete the comment?

themagicianking replied:
Resolved [here](https://github.com/Techtonica/techtonica.org/pull/534/commits/9ff7dcad603d28899b20c30486a3b0c681e31933)!

themagicianking commented on templates/full-time-program.html:
```diff
@@ -5,238 +5,311 @@
-      <p class="justify">
-        Tuition for the 2025 H1 cohort is $18,000. The total cost has been subsidized to lower 
-        the price to an amount that is less expensive than other programs at a cost of $1,000 
-        per week (from Hackbright at $1,075/week to General Assembly at $2,183/week). Tuition 
-        paid upfront receives a 15% overall discount. For example, the full $18,000 cost of 
-        tuition when paid upfront with the 15% discount comes out to $15,300.
-      </p>
-      <p class="justify">
-        If you will not have $18,000 beyond 10 months of living costs, you are eligible to apply 
-        for a partial or full scholarship. If you will not be able to cover tuition or 10 months 
-        of living costs, you are eligible to apply for a tuition scholarship as well as a monthly 
-        stipend. Only those determined to be in need of full scholarships will 
-        be considered for monthly stipends. At present, 90-95% of past applicants would have been 
-        eligible for the needs-based, sliding-scale tuition and stipend scholarships with today’s model.
-      </p>
-      <p class="justify">
-        What bootcamps and Techtonica have in common is the fact that they are
-        intensive training. However, bootcamps can be pretty passive classroom 
-        learning with students and they cost between $1,075 and $2,200 per week 
-        without many scholarship opportunities. Techtonica aims to be more like 
-        the workplace to make the transition into the industry better—instead of 
-        having an instructor constantly guiding you (which you won't have in any 
-        software engineering job), you are given a general schedule of topics and 
-        assignments (see
-        <a title="Curriculum" target="_blank" href="https://github.com/techtonica/curriculum">our curriculum repo</a>
-        for an idea of the resources provided) and you're expected to figure out
-        what to do on your own or in collaboration with volunteers, mentors,
-        tech groups, other participants, and Techtonica staff. Because our
-        resources are open-source, they are continuously being improved and you
-        can even make improvements as you work through them. In the best jobs,
-        there are regular check-ins with very understanding managers and mentors
-        provide regular support, but that is not always the case. Techtonica’s
-        program is as hands-on and project-based as possible to best prepare you
-        for work in tech. You'll also be expected to speak and present in front
-        of people often—a very useful skill to have for a successful career.
-      </p>
-      <p>Watch our 
-        <a title="Tuition Video" target="_blank" href="https://youtu.be/ateYQ7Qb3l8">Techtonica's Sliding Scale, Needs Based Tuition Q&A Session</a>,
-        to learn more.
-      </p>
```
I'm unsure whether I removed this during rebase or changed it manually; in either case, I know there was some error  with this that I was trying to fix wrt that first p tag. Prettier kept giving me the error "unexpected closing tag"; apparently that also happens when a tag contains something that shouldn't be nested inside it (so, the ordered list, in this case). [I removed the p tag and added back in the incorrect deletions with git difftool](https://github.com/Techtonica/techtonica.org/pull/534/commits/aa31951360524850bde74c7f5111128d2707c96a). I believe this shouldn't overwrite any changes in develop but would love a double check.

themagicianking commented on templates/full-time-program.html:
```diff
@@ -5,238 +5,311 @@
       </h2>
     </div>
   </div>
-      <div class="column">
```
See below comment :)

themagicianking commented:
I believe that's every comment resolved! Re-requested reviews for confirmation this is good to merge in?

jsnorek reviewed (approved):
awesome job navigating all the rebase stuff!

themagicianking commented on main_site.py:
```diff
@@ -28,18 +29,34 @@
+    app_open_date = datetime.datetime.strptime(
+        app_open_date_string, "%m/%d/%y %H:%M:%S"
+    )
+except BaseException:
```
@monikkaelyse thank you so much--yes, I would love help with implementation details with this. Would it be better to set defaults or to throw an error in the same way the eventbrite exception does, and what defaults would you recommend setting? It seems like it might be safer to have the default be closed applications in case variables are unset outside of the application window.

daaimah123 commented on main_site.py:
```diff
@@ -28,18 +29,36 @@
+try:
+    app_open_date_string = os.environ["APP_OPEN_DATE"]
+    is_extended = os.environ["APP_EXTENDED"].lower() == "true"
+    print("hi this ran")
```
sure @monikkaelyse, that would be a huge help, thanks for volunteering and being so detail oriented.
